### PR TITLE
Create spiking_neuron.jl

### PR DESCRIPTION
**Pull request by Aadesha**
this is spiking neuron model can be found here, https://flowstar.org/examples/.

### CODE REVIEW

File: spiking_neurons/spiking_neuron.jl [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    local I = 40
+    dx[1] = 0.04*x[1]^2 + 5*x[1] + 140 - x[2] + I
+    dx[2] = a*(b*x[1] - x[2])
+end
+
+function spiking_neurons()
+
+    automaton = LightAutomaton(1) # one node
+
+    inv_one = HPolyhedron([HalfSpace([-1.0, 0.0], -0.2),    #x[1] >= -0.2
```
Comment (schillic) [inline spiking_neurons/spiking_neuron.jl]:
```suggestion
    inv_one = HPolyhedron([HalfSpace([-1.0, 0.0], 0.2),    #x[1] >= -0.2
```
The constraints in this file are wrong. `ax <= b` is equivalent to `-ax >= -b`.

Comment (Aadesha) [reply spiking_neurons/spiking_neuron.jl]: Thanks, now what remains to plot x[1] vs t and [2] vs t graphs.

File: spiking_neurons/spiking_neuron.jl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    inv_one = HPolyhedron([HalfSpace([-1.0, 0.0], 0.2),    #x[1] >= -0.2
+                           HalfSpace([0.0, -1.0], 70.0)])  #x[2] >= -70
+
+    m1 = ConstrainedBlackBoxContinuousSystem(spiking_neuron!, 2, inv_one)
+
+    modes = [m1]
+
+    add_transition!(automaton, 1, 1, 1)
+
+    guard_alpha = HPolyhedron([HalfSpace([1.0, 0.0], 30.0),   # x[1] >= 30
```
Comment (schillic) [inline spiking_neurons/spiking_neuron.jl]:
```suggestion
    guard_alpha = HPolyhedron([HalfSpace([-1.0, 0.0], -30.0),   # x[1] >= 30
```

File: spiking_neurons/spiking_neuron.jl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                               HalfSpace([-1.0, 0.0], -30.0)]) # x[1] <= 30
```
Comment (schillic) [inline spiking_neurons/spiking_neuron.jl]:
```suggestion
                               HalfSpace([1.0, 0.0], 30.0)]) # x[1] <= 30
```

File: spiking_neurons/spiking_neuron.jl [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    options = Options(:mode=>"reach", :T=>100.0, :plot_vars=>[1, 2], :project_reachset=>false)
+
+    return (system, options)
+end
+
+problem, options = spiking_neurons();
+
+options = Options(:mode=>"reach", :T=>100.0, :plot_vars=>[1, 2], :project_reachset=>false, :verbosity => "info")
+
+@time sol_TMJets = solve(problem, options, TMJets(:orderT=>5, :orderQ=>2, :abs_tol=>1e-10),LazyDiscretePost(:check_invariant_intersection=>true))
```
Comment (schillic) [inline spiking_neurons/spiking_neuron.jl]:
```suggestion
opC = TMJets(:orderT=>5, :orderQ=>2, :abs_tol=>1e-10)
opD = LazyDiscretePost(:check_invariant_intersection=>true)
@time sol_TMJets = solve(problem, options, opC, opD)
```

**Comment by mforets** [comment]
The model `spiking_neurons()` looks good to me, please add at least a reference to https://flowstar.org/examples/.

File: spiking_neurons/spiking_neuron.jl [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+using Reachability: solve
+using Plots
+using TaylorIntegration
+
+@taylorize function spiking_neuron!(t, x, dx)
+    local a = 0.02
+    local b = 0.2
+    local I = 40
+    dx[1] = 0.04*x[1]^2 + 5*x[1] + 140 - x[2] + I
+    dx[2] = a*(b*x[1] - x[2])
```
Comment (schillic) [inline spiking_neurons/spiking_neuron.jl]: IIRC (but I might be wrong), there are performance issues with nested multiplications. Maybe the following is better:
```suggestion
    dx[2] = a*b*x[1] - a*x[2]
```
@mforets: Do you know?

Comment (mforets) [reply spiking_neurons/spiking_neuron.jl]: use binary operations with `tayorize`, so i think that here `a*(b*x[1] - x[2])` is ok

Comment (mforets) [reply spiking_neurons/spiking_neuron.jl]: however, the differential equation for `x[1]` should be rewritten eg.

```julia
dx[1] = (0.04*(x[1]*x[1]) + 5*x[1]) + (140+I-x[2])
```

**Comment by Aadesha** [comment]
Thanks I will make changes:)

File: spiking_neurons/spiking_neuron.jl (hunk 5 above)
Comment (lbenet) [reply spiking_neurons/spiking_neuron.jl]: With regards to using `@taylorize`, Marcelo's solution is ok; another option is
```julia
dx[1] = (0.04*x[1]^2) + 5*x[1]) + ((140+I)-x[2])
```
The important thing is that you use parenthesis when more than two operations/functions are involved. This restriction is related to the way the internals of `@taylorize` work.

Comment (schillic) [reply spiking_neurons/spiking_neuron.jl]: One parenthesis is unmatched.

File: spiking_neurons/spiking_neuron.jl [hunk 6]
```diff
@@ -0,0 +1,77 @@
+using Revise, MathematicalSystems,Reachability, LinearAlgebra, HybridSystems
+using LazySets;
+using Reachability: solve
+using Plots
+using TaylorIntegration
+
+@taylorize function spiking_neuron!(t, x, dx)
```
Comment (lbenet) [inline spiking_neurons/spiking_neuron.jl]: What version of `TaylorIntegration.jl` is used? My guess is a version < v0.5.0, which would then be compatible to the order of the arguments of `spiking_neuron!(...)`; if version v0.5.0 is used, the parameters should read `spiking_neuron!(dx, x, params, t)`.

Comment (mforets) [reply spiking_neurons/spiking_neuron.jl]: Indeed, it is using a version < v0.5.0. We'll get to issue https://github.com/JuliaReach/Reachability.jl/issues/651 in Reachability next week.

File: spiking_neurons/spiking_neuron.jl [hunk 7]
```diff
@@ -0,0 +1,82 @@
+#=
+Model : spiking neuron
+model can be formalized by a hybrid automaton consists of 2 modes and 2 variables.
```
Comment (mforets) [inline spiking_neurons/spiking_neuron.jl]:
```suggestion
model can be formalized by a hybrid automaton consists of 1 mode and 2 variables.
```

Comment (Aadesha) [reply spiking_neurons/spiking_neuron.jl]: Yes it have one mode

**Pull request merged**